cli/crates/backend/src/project.rs

```diff
@@ -19,40 +33,83 @@ use std::path::PathBuf;
 /// - returns [`BackendError::CreateGrafbaseDirectory`] if the grafbase directory cannot be created
 ///
 /// - returns [`BackendError::WriteSchema`] if the schema file cannot be written
-pub fn init(name: Option<&str>) -> Result<(), BackendError> {
+///
+/// ## Templates
+///
+/// - returns [`BackendError::UnsupportedTemplateURL`] if a template URL is not supported
+///
+/// - returns [`BackendError::UnsupportedTemplate`] if a template name is not supported or contains unsupported characters
+///
+/// - returns [`BackendError::StartDownloadRepoArchive`] if a template URL is not supported (if the request could not be made)
+///
+/// - returns [`BackendError::DownloadRepoArchive`] if a repo tar could not be downloaded (on a non 200-299 status)
+///
+/// - returns [`BackendError::StoreRepoArchive`] if a repo tar could not be stored
+///
+/// - returns [`BackendError::NoFilesExtracted`] if no files were extracted from the repo archive
+///
+/// - returns [`BackendError::MoveExtractedFiles`] if the extracted files from the template repository could not be moved
+///
+/// - returns [`BackendError::ReadArchiveEntries`] if the entries of the template repository archive could not be read
+///
+/// - returns [`BackendError::ExtractArchiveEntry`] if one of the entries of the template repository archive could not be extracted
+///
+/// - returns [`BackendError::CleanExtractedFiles`] if the files extracted from the template repository archive could not be cleaned
+pub fn init(name: Option<&str>, template: Option<&str>) -> Result<(), BackendError> {
     let project_path = to_project_path(name)?;
     let grafbase_path = project_path.join(GRAFBASE_DIRECTORY);
     let schema_path = grafbase_path.join(GRAFBASE_SCHEMA);
 
-    if schema_path.exists() {
-        Err(BackendError::AlreadyAProject(schema_path))
+    if grafbase_path.exists() {
+        Err(BackendError::AlreadyAProject(grafbase_path))
+    } else if let Some(template) = template {
+        if let Ok(repo_url) = Url::parse(template) {
```

Comment:
As discussed, I think we should flip the logic to first probe for a simple template name (`[a-zA-Z-]+`) and then proceed to treating the input as an URL as a fallback. 🙂

Reply:
I'm debating whether this would be a bit limiting, we can't assume templates will only use alphanumeric characters and dashes

Reply:
IMO it's definitely better than what we have currently. If you pass something URL-like that fails to parse, it should result in an appropriate error.

Reply:
Even something like:
```rust
if template.starts_with("https://") {
    // Treat as an URL.
} else {
    // Treat as a template name.
}
```

would be better than this "fall back on error" thing.

Reply:
Semantically it's "fall back if the input is not a proper URL", which is more or less the same as you're doing with the protocol check

Reply:
As in - All cases of urls without `http://` and `https://` will currently be treated as a template name, exactly like the example

Reply:
Regardless we can add a check for the specific scheme

Reply:
URL parsing can fail in various other ways, so no. “Does it start with a particular scheme” is a heuristic check for how to treat the input, and then whether that input is valid in light of the distinction made is determined separately. How to treat the input isn't determined by the initial guess of what it is, in other words. Okay, this isn't important enough to be arguing over too much, so I won't be insistent over this if you're set on doing it this way. That said, you do have a FIXME related to it, so later it'll need to be addressed one way or another.

Reply:
I don't see this as an argument, this is a discussion. My goal isn't to advocate for this solution specifically, it's to find one that would be robust.

I'm not particularly married to checking if it's a URL first, my goal is to avoid cases where we'd need to redeploy because we used a type of character we didn't account for in a template name (e.g. `todo-2`).

My point before is that checking for the protocol wouldn't add anything extra since it's already being done by the parser, if your goal is presenting a better error message in the case of an invalid URL with a scheme we could do this check first, but that'd rule out trying to append a scheme if one doesn't exist which we also discussed earlier (although that'd also have some edge cases in case of templates containing dots, which is why it's not included here)

The `FIXME` isn't specifically related to this issue IMO, it derives from knowing whether a template name is correct only after downloading the archive, for which we'll need to check the archive in advance (pre-extraction) and tweak the error message to cover both cases.

All the above being said I still think we should consider separate arguments for URLs and template names since they'd rule these cases out and be more discoverable (with not much of a DX price), but given what we're doing currently there's tradeoffs to any sort of heuristic we choose

Reply:
Since we're streaming the tar, checking the contents before extraction is tricky. 

We could treat the absence of extracted files as definite evidence that the files didn't exist in the tar but there are other cases that could cause that.

We could set a variable after first extracting a file but that isn't as clean as I would like

Reply:
I didn't mean “argument” in a negative sense of the word. 🙂 Nothing wrong with civil arguments. 🙂  I meant to say that I would rather we kept iterating on this, hey, even making mistakes and perhaps improving in future versions rather than spending too much time on overthinking this.

My general point is this: this is tying the initial guess of “is this an URL or is this template name” to the validation informed by one of the two possible outcomes. To me it doesn't seem right on a purely logical basis, but incidentally it also causes inaccurate error reporting in a particular scenario. The FIXME seems related because that branch of code will specifically kick in for the case of an invalid URL, and that's what would be avoided by decoupling these two concerns.

Anyway, as I said, not a blocker.

Reply:
I see, I actually added a check for `/` at some point in the template name since URLs would almost always contain a slash and directories cannot have one, so most of these cases don't reach the `FIXME` anymore, hence me not associating it with this part of the code.

The one case that'd still trigger it is something like `mysite.com` which we wouldn't cover by checking the scheme (but also can't be ruled out as a template name)

Checking for a scheme would provide a better error message in a case like `https://%` though because it would consider it a URL rather than an invalid template name (before downloading the archive). Although we could also just consider anything with a slash as a malformed URL due to templates not being able to have a slash at all

Reply:
@yoav-lavi oh, yes, that's much better, thanks! 🙂

Reply:
@jakubadamw could you please re-review?